source/retryable-writes/retryable-writes.rst

```diff
@@ -181,59 +181,94 @@ Retrying Write Commands
 Drivers MUST only attempt to retry a write command that encounters a retryable
 error. Drivers MUST NOT attempt to retry a write command on any other error.
 
+When selecting a writable server for the first attempt of a retryable write
+command, drivers MUST allow a server selection error to propagate. If the
+selected server's maximum wire version does not support retryable writes,
+drivers MUST raise a client-side error. In both cases, the caller is able to
+infer that no attempt was made.
+
 If the first attempt of a write command including a transaction ID encounters a
 retryable error, the driver MUST update its topology according to the SDAM spec
-(see: `Error Handling`_), reselect a writable server, and execute the command
-again. Consider the following pseudo-code:
+(see: `Error Handling`_) and capture this original retryable error. Drivers
+should then proceed with selecting a writable server for the retry attempt.
 
 .. _Error Handling: ../server-discovery-and-monitoring/server-discovery-and-monitoring.rst#error-handling
 
+If the driver cannot select a server for the retry attempt or the selected
+server does not support retryable writes (e.g. mixed-version cluster), retrying
+is not possible and drivers MUST raise the original retryable error. In both
+cases, the caller is able to infer that an attempt was made.
+
+If the retry attempt also fails, drivers MUST update their topology according to
+the SDAM spec (see: `Error Handling`_) and raise any ensuing error to the user.
+In this case, the error from the retry attempt is just as relevant as the
+original retryable error and the caller is still able to infer that an attempt
+was made.
+
+Drivers MUST NOT attempt to retry a write command with the same transaction ID
+more than once.
+
+Consider the following pseudo-code:
```

Comment:
I moved most of the prose describing the retryable algorithm above the code example, as I felt the spec language is the primary source of truth and any pseudo-code should merely follow up and illustrate that language.

Reply:
One further issue to address: what if server selection succeeds for the driver fails to either:

* get a connection from the pool (e.g. pool exhaustion)
* open a new connection (connect error)

By the same argument, in both those cases the original error should be thrown.

Reply:
Are those exceptions that would both arise from `executeCommand(server, command);`? The spec does not consider them to be retryable errors, so they wouldn't be caught by the first attempt and would immediately propagate to the calling context.

Are you suggesting we catch these in the final `return executeCommand(server, command);` for the retry attempt, so that we can throw the original error?

Reply:
Yes, that's what I'm suggesting. Because like server selection they're not retryable error it makes sense to throw the original error for them as well.

Reply:
I revised the pseudo-code in the most recent change. I also demonstrated that we update SDAM for a NetworkException or NotMasterException on the second attempt.